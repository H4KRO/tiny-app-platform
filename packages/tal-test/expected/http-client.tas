set_system_property(key: "title", value: "HTTP Client demo")
var headers = []
var body = null
var language = "text"
var doFormat = false
var doShowHeaders = false
var method = "GET"
var url
var response
var responseBodyString
fun executeHttpRequest() {
  set response = http_request(
    method
    url
    headers | array_filter(it => it[0] && it[1])
    body
    allowErrorStatusCode: true
  )
  try {
    let contentType = string_split(response.headers["content-type"], ";")[0] | string_trim
    if (contentType == "text/html") {
      set language = "html"
    } else if (contentType == "application/json") {
      set language = "json"
      set doFormat = true
    } else if (contentType == "text/javascript") {
      set language = "javascript"
    }
  }
  set responseBodyString = bytes_to_string(response.body)
}

Column {
  Row {
    Box {
      flex: 0

      Select {
        bindTo:      method
        placeholder: "Method"
        options:     [
          "GET"
          "POST"
          "PUT"
          "PATCH"
          "DELETE"
          "LIST"
          "HEAD"
          "OPTIONS"
        ]
      }
    }

    Box {
      flex: 1

      InputText {
        bindTo:      url
        placeholder: "URL"
        type:        "url"
        onSubmit:    executeHttpRequest
      }
    }
  }

  array_map(
    headers
    (it, index) => [
      Row {
        Box {
          flex: 1

          InputText {
            placeholder: "Header name"
            bindTo:      it[0]
          }
        }

        Box {
          flex: 3

          InputText {
            placeholder: "Header value"
            bindTo:      it[1]
          }
        }

        Box {
          flex: 0

          Button {
            text:    "Remove"
            onClick: () => set headers = array_remove(headers, index)
          }
        }
      }
    ]
  )

  Button {
    onClick: () => set headers = array_append(
      headers
      [
        "user-agent"
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/119.0"
      ]
    )
    text:    "Add user agent header"
  }

  Button {
    text:    "New header"
    onClick: { () => set headers = array_append(headers, ["", ""]) }
  }

  InputText {
    bindTo:      body
    placeholder: "body"
    multiline:   true
  }

  Button {
    text:    "Execute HTTP Request"
    onClick: executeHttpRequest
  }

  {
    let validHeaders = headers | array_filter(it => it[0] && it[1])

    [
      if (method != null && url != null) {
        Text {
          copy: true
          text: "curl -X $method $headers '$url'"
             | string_format({
                method:  method
                url:     url
                headers: validHeaders
                   | array_map(it => string_format(
                      "-H '$name: $value'"
                      {name: it[0], value: it[1]}
                    ))
                   | array_join(" ")
              })
        }
      }

      if (method != null && url != null) {
        Text {
          copy: true
          text: 'http_request(method: "$method", url: "$url" $headers $body)'
             | string_format({
                method:  method
                url:     url
                body:    if (body != null && body != "") {
                  ', body: "' + body + '"'
                } else {
                  " "
                }
                headers: if (array_length(validHeaders) > 0) {
                  ", headers:" + (validHeaders | json_stringify)
                } else {
                  " "
                }
              })
        }
      }
    ]
  }

  if (response != null) {
    Column {
      Row {
        Text {
          text: string_format(
            "Status code: $status"
            {status: response.status}
          )
        }
      }

      Row {
        Text { text: "Show headers" }
        Switch { bindTo: doShowHeaders }
      }

      if (doShowHeaders) {
        Column {
          flexShrink: 0

          Text { text: "Headers:" }

          Table {
            columns: [
              {
                description: "Name"
                display:     row => Text { text: row.key, copy: true }
              }

              {
                description: "Value"
                display:     row => Text { text: row.value, copy: true }
              }
            ]
            values:  response.headers | object_entries | array_map(it => {key: it[0], value: it[1]})
          }
        }
      }

      Row {
        Text { text: "Language syntax to highlight" }

        Select {
          options: [
            "text"
            "json"
            "html"
            "xml"
            "javascript"
            "css"
          ]
          bindTo:  language
        }

        Text { text: "Format code (when possible)" }
        Switch { bindTo: doFormat }
      }

      Box {
        scroll: true

        if (language == "json" && doFormat) {
          Debug {
            value: responseBodyString | json_parse
            force: true
          }
        } else {
          Snippet {
            text:     responseBodyString
            language: language
            format:   doFormat
          }
        }
      }
    }
  }
}
