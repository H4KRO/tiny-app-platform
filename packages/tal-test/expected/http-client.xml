<program>
  <call>
    <local />
    <literal />
    <literal />
  </call>
  <declarelocal>
    <array>
    </array>
  </declarelocal>
  <declarelocal>
    <literal />
  </declarelocal>
  <declarelocal>
    <literal />
  </declarelocal>
  <declarelocal>
    <literal />
  </declarelocal>
  <declarelocal>
    <literal />
  </declarelocal>
  <declarelocal>
    <literal />
  </declarelocal>
  <declarelocal>
  </declarelocal>
  <declarelocal>
  </declarelocal>
  <declarelocal>
  </declarelocal>
  <declarelocal>
    <function>
      <block>
        <assign>
          <local />
          <call>
            <local />
            <local />
            <local />
            <pipe>
              <local />
              <call>
                <local />
                <function>
                  <binaryoperator>
                    <index>
                      <literal />
                      <local />
                    </index>
                    <index>
                      <literal />
                      <local />
                    </index>
                  </binaryoperator>
                </function>
              </call>
            </pipe>
            <local />
            <literal />
          </call>
        </assign>
        <try>
          <block>
            <declarelocal>
              <pipe>
                <index>
                  <literal />
                  <call>
                    <local />
                    <index>
                      <literal />
                      <attribute>
                        <local />
                      </attribute>
                    </index>
                    <literal />
                  </call>
                </index>
                <local />
              </pipe>
            </declarelocal>
            <if>
              <binaryoperator>
                <local />
                <literal />
              </binaryoperator>
              <block>
                <assign>
                  <local />
                  <literal />
                </assign>
              </block>
              <if>
                <binaryoperator>
                  <local />
                  <literal />
                </binaryoperator>
                <block>
                  <assign>
                    <local />
                    <literal />
                  </assign>
                  <assign>
                    <local />
                    <literal />
                  </assign>
                </block>
                <if>
                  <binaryoperator>
                    <local />
                    <literal />
                  </binaryoperator>
                  <block>
                    <assign>
                      <local />
                      <literal />
                    </assign>
                  </block>
                </if>
              </if>
            </if>
          </block>
        </try>
        <assign>
          <local />
          <call>
            <local />
            <attribute>
              <local />
            </attribute>
          </call>
        </assign>
      </block>
    </function>
  </declarelocal>
  <kindedrecord>
    <local />
    <kindedrecord>
      <local />
      <kindedrecord>
        <local />
        <kindedrecord>
          <local />
          <kindedrecordentry>
            <local />
          </kindedrecordentry>
          <kindedrecordentry>
            <literal />
          </kindedrecordentry>
          <kindedrecordentry>
            <array>
              <literal />
              <literal />
              <literal />
              <literal />
              <literal />
              <literal />
              <literal />
              <literal />
            </array>
          </kindedrecordentry>
        </kindedrecord>
        <kindedrecordentry>
          <literal />
        </kindedrecordentry>
      </kindedrecord>
      <kindedrecord>
        <local />
        <kindedrecord>
          <local />
          <kindedrecordentry>
            <local />
          </kindedrecordentry>
          <kindedrecordentry>
            <literal />
          </kindedrecordentry>
          <kindedrecordentry>
            <literal />
          </kindedrecordentry>
          <kindedrecordentry>
            <local />
          </kindedrecordentry>
        </kindedrecord>
        <kindedrecordentry>
          <literal />
        </kindedrecordentry>
      </kindedrecord>
    </kindedrecord>
    <call>
      <local />
      <local />
      <function>
        <array>
          <kindedrecord>
            <local />
            <kindedrecord>
              <local />
              <kindedrecord>
                <local />
                <kindedrecordentry>
                  <literal />
                </kindedrecordentry>
                <kindedrecordentry>
                  <index>
                    <literal />
                    <local />
                  </index>
                </kindedrecordentry>
              </kindedrecord>
              <kindedrecordentry>
                <literal />
              </kindedrecordentry>
            </kindedrecord>
            <kindedrecord>
              <local />
              <kindedrecord>
                <local />
                <kindedrecordentry>
                  <literal />
                </kindedrecordentry>
                <kindedrecordentry>
                  <index>
                    <literal />
                    <local />
                  </index>
                </kindedrecordentry>
              </kindedrecord>
              <kindedrecordentry>
                <literal />
              </kindedrecordentry>
            </kindedrecord>
            <kindedrecord>
              <local />
              <kindedrecord>
                <local />
                <kindedrecordentry>
                  <literal />
                </kindedrecordentry>
                <kindedrecordentry>
                  <function>
                    <assign>
                      <local />
                      <call>
                        <local />
                        <local />
                        <local />
                      </call>
                    </assign>
                  </function>
                </kindedrecordentry>
              </kindedrecord>
              <kindedrecordentry>
                <literal />
              </kindedrecordentry>
            </kindedrecord>
          </kindedrecord>
        </array>
      </function>
    </call>
    <kindedrecord>
      <local />
      <kindedrecordentry>
        <function>
          <assign>
            <local />
            <call>
              <local />
              <local />
              <array>
                <literal />
                <literal />
              </array>
            </call>
          </assign>
        </function>
      </kindedrecordentry>
      <kindedrecordentry>
        <literal />
      </kindedrecordentry>
    </kindedrecord>
    <kindedrecord>
      <local />
      <kindedrecordentry>
        <literal />
      </kindedrecordentry>
      <kindedrecordentry>
        <block>
          <function>
            <assign>
              <local />
              <call>
                <local />
                <local />
                <array>
                  <literal />
                  <literal />
                </array>
              </call>
            </assign>
          </function>
        </block>
      </kindedrecordentry>
    </kindedrecord>
    <kindedrecord>
      <local />
      <kindedrecordentry>
        <local />
      </kindedrecordentry>
      <kindedrecordentry>
        <literal />
      </kindedrecordentry>
      <kindedrecordentry>
        <literal />
      </kindedrecordentry>
    </kindedrecord>
    <kindedrecord>
      <local />
      <kindedrecordentry>
        <literal />
      </kindedrecordentry>
      <kindedrecordentry>
        <local />
      </kindedrecordentry>
    </kindedrecord>
    <block>
      <declarelocal>
        <pipe>
          <local />
          <call>
            <local />
            <function>
              <binaryoperator>
                <index>
                  <literal />
                  <local />
                </index>
                <index>
                  <literal />
                  <local />
                </index>
              </binaryoperator>
            </function>
          </call>
        </pipe>
      </declarelocal>
      <array>
        <if>
          <binaryoperator>
            <binaryoperator>
              <local />
              <literal />
            </binaryoperator>
            <binaryoperator>
              <local />
              <literal />
            </binaryoperator>
          </binaryoperator>
          <block>
            <kindedrecord>
              <local />
              <kindedrecordentry>
                <literal />
              </kindedrecordentry>
              <kindedrecordentry>
                <pipe>
                  <literal />
                  <call>
                    <local />
                    <record>
                      <recordentry>
                        <local />
                      </recordentry>
                      <recordentry>
                        <local />
                      </recordentry>
                      <recordentry>
                        <pipe>
                          <local />
                          <call>
                            <local />
                            <function>
                              <call>
                                <local />
                                <literal />
                                <record>
                                  <recordentry>
                                    <index>
                                      <literal />
                                      <local />
                                    </index>
                                  </recordentry>
                                  <recordentry>
                                    <index>
                                      <literal />
                                      <local />
                                    </index>
                                  </recordentry>
                                </record>
                              </call>
                            </function>
                          </call>
                          <call>
                            <local />
                            <literal />
                          </call>
                        </pipe>
                      </recordentry>
                    </record>
                  </call>
                </pipe>
              </kindedrecordentry>
            </kindedrecord>
          </block>
        </if>
        <if>
          <binaryoperator>
            <binaryoperator>
              <local />
              <literal />
            </binaryoperator>
            <binaryoperator>
              <local />
              <literal />
            </binaryoperator>
          </binaryoperator>
          <block>
            <kindedrecord>
              <local />
              <kindedrecordentry>
                <literal />
              </kindedrecordentry>
              <kindedrecordentry>
                <pipe>
                  <literal />
                  <call>
                    <local />
                    <record>
                      <recordentry>
                        <local />
                      </recordentry>
                      <recordentry>
                        <local />
                      </recordentry>
                      <recordentry>
                        <if>
                          <binaryoperator>
                            <binaryoperator>
                              <local />
                              <literal />
                            </binaryoperator>
                            <binaryoperator>
                              <local />
                              <literal />
                            </binaryoperator>
                          </binaryoperator>
                          <block>
                            <binaryoperator>
                              <binaryoperator>
                                <literal />
                                <local />
                              </binaryoperator>
                              <literal />
                            </binaryoperator>
                          </block>
                          <block>
                            <literal />
                          </block>
                        </if>
                      </recordentry>
                      <recordentry>
                        <if>
                          <binaryoperator>
                            <call>
                              <local />
                              <local />
                            </call>
                            <literal />
                          </binaryoperator>
                          <block>
                            <binaryoperator>
                              <literal />
                              <nested>
                                <pipe>
                                  <local />
                                  <local />
                                </pipe>
                              </nested>
                            </binaryoperator>
                          </block>
                          <block>
                            <literal />
                          </block>
                        </if>
                      </recordentry>
                    </record>
                  </call>
                </pipe>
              </kindedrecordentry>
            </kindedrecord>
          </block>
        </if>
      </array>
    </block>
    <if>
      <binaryoperator>
        <local />
        <literal />
      </binaryoperator>
      <block>
        <kindedrecord>
          <local />
          <kindedrecord>
            <local />
            <kindedrecord>
              <local />
              <kindedrecordentry>
                <call>
                  <local />
                  <literal />
                  <record>
                    <recordentry>
                      <attribute>
                        <local />
                      </attribute>
                    </recordentry>
                  </record>
                </call>
              </kindedrecordentry>
            </kindedrecord>
          </kindedrecord>
          <kindedrecord>
            <local />
            <kindedrecord>
              <local />
              <kindedrecordentry>
                <literal />
              </kindedrecordentry>
            </kindedrecord>
            <kindedrecord>
              <local />
              <kindedrecordentry>
                <local />
              </kindedrecordentry>
            </kindedrecord>
          </kindedrecord>
          <if>
            <local />
            <block>
              <kindedrecord>
                <local />
                <kindedrecord>
                  <local />
                  <kindedrecordentry>
                    <literal />
                  </kindedrecordentry>
                </kindedrecord>
                <kindedrecord>
                  <local />
                  <kindedrecordentry>
                    <array>
                      <record>
                        <recordentry>
                          <literal />
                        </recordentry>
                        <recordentry>
                          <function>
                            <kindedrecord>
                              <local />
                              <kindedrecordentry>
                                <attribute>
                                  <local />
                                </attribute>
                              </kindedrecordentry>
                              <kindedrecordentry>
                                <literal />
                              </kindedrecordentry>
                            </kindedrecord>
                          </function>
                        </recordentry>
                      </record>
                      <record>
                        <recordentry>
                          <literal />
                        </recordentry>
                        <recordentry>
                          <function>
                            <kindedrecord>
                              <local />
                              <kindedrecordentry>
                                <attribute>
                                  <local />
                                </attribute>
                              </kindedrecordentry>
                              <kindedrecordentry>
                                <literal />
                              </kindedrecordentry>
                            </kindedrecord>
                          </function>
                        </recordentry>
                      </record>
                    </array>
                  </kindedrecordentry>
                  <kindedrecordentry>
                    <pipe>
                      <attribute>
                        <local />
                      </attribute>
                      <local />
                      <call>
                        <local />
                        <function>
                          <record>
                            <recordentry>
                              <index>
                                <literal />
                                <local />
                              </index>
                            </recordentry>
                            <recordentry>
                              <index>
                                <literal />
                                <local />
                              </index>
                            </recordentry>
                          </record>
                        </function>
                      </call>
                    </pipe>
                  </kindedrecordentry>
                </kindedrecord>
                <kindedrecordentry>
                  <literal />
                </kindedrecordentry>
              </kindedrecord>
            </block>
          </if>
          <kindedrecord>
            <local />
            <kindedrecord>
              <local />
              <kindedrecordentry>
                <literal />
              </kindedrecordentry>
            </kindedrecord>
            <kindedrecord>
              <local />
              <kindedrecordentry>
                <array>
                  <literal />
                  <literal />
                  <literal />
                  <literal />
                  <literal />
                  <literal />
                </array>
              </kindedrecordentry>
              <kindedrecordentry>
                <local />
              </kindedrecordentry>
            </kindedrecord>
            <kindedrecord>
              <local />
              <kindedrecordentry>
                <literal />
              </kindedrecordentry>
            </kindedrecord>
            <kindedrecord>
              <local />
              <kindedrecordentry>
                <local />
              </kindedrecordentry>
            </kindedrecord>
          </kindedrecord>
          <kindedrecord>
            <local />
            <if>
              <binaryoperator>
                <binaryoperator>
                  <local />
                  <literal />
                </binaryoperator>
                <local />
              </binaryoperator>
              <block>
                <kindedrecord>
                  <local />
                  <kindedrecordentry>
                    <pipe>
                      <local />
                      <local />
                    </pipe>
                  </kindedrecordentry>
                  <kindedrecordentry>
                    <literal />
                  </kindedrecordentry>
                </kindedrecord>
              </block>
              <block>
                <kindedrecord>
                  <local />
                  <kindedrecordentry>
                    <local />
                  </kindedrecordentry>
                  <kindedrecordentry>
                    <local />
                  </kindedrecordentry>
                  <kindedrecordentry>
                    <local />
                  </kindedrecordentry>
                </kindedrecord>
              </block>
            </if>
            <kindedrecordentry>
              <literal />
            </kindedrecordentry>
          </kindedrecord>
        </kindedrecord>
      </block>
    </if>
  </kindedrecord>
</program>
